pkg/operator/osd/osd.go

```diff
@@ -289,7 +289,9 @@ func (c *Cluster) osdContainer(devices []Device, directories []Directory, select
 		privileged = true
 	}
 	runAsUser := int64(0)
-	runAsNonRoot := false
+	// don't set runAsNonRoot explicitly when it is false, Kubernetes version < 1.6.4 has
+	// an issue with this fixed in https://github.com/kubernetes/kubernetes/pull/47009
+	// runAsNonRoot := false
```

Comment:
what is the repercussion of not setting the `runAsNonRoot` in newer version K8S, like 1.8? Will it set the SecurityContext to run as Root?

Reply:
Not setting it just means there will be no validation on whether it is running as root or not.

From https://kubernetes.io/docs/api-reference/v1.8/#securitycontext-v1-core

Indicates that the container must run as a non-root user. If true, the Kubelet will validate the image at runtime to ensure that it does not run as UID 0 (root) and fail to start the container if it does. **If unset or false, no such validation will be performed.**

The problem is with for environment with more restricted PodSecurityPolicy, where a user or service account may have permissions to use multiple PSP, and without explicit securityContext the system may only USE a policy that is not sufficient.   For example, a user may have access to two PSPs, one that can only runAsNonRoot: false so you can't runAsUser:0 , and another PSP that can runAsNonRoot: true.  When runAsNonRoot is missing in the pod template, my observation is that both PSP qualify to run the pod, but if it picks the wrong one the pod would fail to run.  I don't know how it actually picks that, but we have seen confusions from our users running into this issue.

We can argue that it is probably a misconfiguration that a user/service account can have access to multiple PSPs that "conflicts", but I am not sure what the right thing to do here is, and we have learned to be very explicit about securityContext to avoid these issues.

Reply:
Cool. Thanks @phsiao